src/parsing/prscParser.ts

```diff
@@ -0,0 +1,2795 @@
+				preceded(surrounded(token('='), whitespacePlus), stringLiteral),
+				(namePart, value) => [
+					'decimalFormatParam',
+					['decimalFormatParamName', namePart],
+					['decimalFormatParamValue', value],
+				]
+			)
+		),
+		(decimalFormatName, decimalFormatParams) => [
+			'decimalFormatDecl',
+			...(decimalFormatName ? [decimalFormatName] : []),
+			...decimalFormatParams,
+		]
+	);
+
+	const setter: Parser<IAST> = or([
+		boundarySpaceDecl,
+		defaultCollationDecl,
+		baseUriDecl,
+		constructionDecl,
+		orderingModeDecl,
+		emptyOrderDecl,
+		copyNamespacesDecl,
+		decimalFormatDecl,
+	]);
+
+	const optionDecl: Parser<IAST> = then(
+		precededMultiple(
+			[token('declare'), whitespacePlus, token('option'), whitespacePlus],
+			eqName
+		),
+		preceded(whitespacePlus, stringLiteral),
+		(optionName, optionContents) =>
+			['optionDecl', ['optionName', optionName], ['optionContents', optionContents]] as IAST
+	);
+
+	const contextItemDecl: Parser<IAST> = then(
+		precededMultiple(
+			[token('declare'), whitespacePlus, token('context'), whitespacePlus, token('item')],
+			optional(precededMultiple([whitespacePlus, token('as')], itemType))
+		),
+		or([
+			map(preceded(surrounded(token(':='), whitespace), varValue), (x) => ['varValue', x]),
+			map(
+				precededMultiple(
+					[whitespacePlus, token('external')],
+					optional(preceded(surrounded(token(':='), whitespace), varDefaultValue))
+				),
+				(_) => ['external']
+			),
+		]),
+		(itemTypePart, val) =>
+			[
+				'contextItemDecl',
+				...(itemTypePart ? [['contextItemType', itemTypePart]] : []),
+				val,
+			] as IAST
+	);
+
+	const prolog: Parser<IAST> = then(
+		star(
+			followed(
+				or([defaultNamespaceDecl, setter, namespaceDecl, importExpr]),
+				surrounded(separator, whitespace)
+			)
+		),
+		star(
+			followed(
+				or([contextItemDecl, annotatedDecl, optionDecl]),
+				surrounded(separator, whitespace)
+			)
+		),
+		(moduleSettings, declarations) =>
+			moduleSettings.length === 0 && declarations.length === 0
+				? null
+				: ['prolog', ...moduleSettings, ...declarations]
+	);
+
+	const moduleDecl: Parser<IAST> = precededMultiple(
+		[token('module'), whitespacePlus, token('namespace'), whitespacePlus],
+		then(
+			followed(ncName, surrounded(token('='), whitespace)),
+			followed(uriLiteral, surrounded(separator, whitespace)),
+			(prefixPart, uri) => ['moduleDecl', ['prefix', prefixPart], ['uri', uri]]
+		)
+	);
+
+	const libraryModule: Parser<IAST> = then(
+		moduleDecl,
+		preceded(whitespace, prolog),
+		(moduleDeclPart, prologPart) =>
+			['libraryModule', moduleDeclPart, ...(prologPart ? [prologPart] : [])] as IAST
+	);
+
+	const mainModule: Parser<IAST> = then(
+		prolog,
+		preceded(whitespace, queryBody),
+		(prologPart, body) => ['mainModule', ...(prologPart ? [prologPart] : []), body]
+	);
+
+	const versionDeclaration: Parser<IAST> = map(
+		precededMultiple(
+			[token('xquery'), whitespace],
```

Comment:
It might be cool to deduplicate these tokens: add something (maybe even another file to make this one a bit smaller?) that defines them: `const XQUERY_TOKEN = token('xquery'); const BRACE_OPEN_TOKEN = token('(');` etcetera?